worker/src/RTC/ActiveSpeakerObserver.cpp

```diff
@@ -0,0 +1,564 @@
+	{
+		MS_TRACE();
+
+		auto it = this->mapProducerSpeaker.find(producer->id);
+
+		if (it == this->mapProducerSpeaker.end())
+		{
+			return;
+		}
+
+		if (it->second.speaker != nullptr)
+		{
+			delete it->second.speaker;
+		}
+
+		this->mapProducerSpeaker.erase(producer->id);
+
+		if (producer->id == this->dominantId)
+		{
+			Update();
+		}
+	}
+
+	void ActiveSpeakerObserver::ProducerResumed(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		auto it = this->mapProducerSpeaker.find(producer->id);
+
+		if (it != this->mapProducerSpeaker.end())
+		{
+			auto& rtpObserver = it->second;
+
+			rtpObserver.speaker->paused = false;
+		}
+	}
+
+	void ActiveSpeakerObserver::ProducerPaused(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		auto it = this->mapProducerSpeaker.find(producer->id);
+
+		if (it != this->mapProducerSpeaker.end())
+		{
+			auto& rtpObserver = it->second;
+
+			rtpObserver.speaker->paused = true;
+		}
+	}
+
+	void ActiveSpeakerObserver::ReceiveRtpPacket(RTC::Producer* producer, RTC::RtpPacket* packet)
+	{
+		MS_TRACE();
+
+		if (IsPaused())
+			return;
+
+		uint8_t volume;
+		bool voice;
+
+		if (!packet->ReadSsrcAudioLevel(volume, voice))
+			return;
+
+		auto it = this->mapProducerSpeaker.find(producer->id);
+
+		if (it != this->mapProducerSpeaker.end())
+		{
+			auto& rtpObserver = it->second;
+			uint64_t now      = DepLibUV::GetTimeMs();
+
+			rtpObserver.speaker->LevelChanged(volume, now);
+		}
+	}
+
+	void ActiveSpeakerObserver::Paused()
+	{
+		MS_TRACE();
+
+		this->periodicTimer->Stop();
+	}
+
+	void ActiveSpeakerObserver::Resumed()
+	{
+		MS_TRACE();
+
+		this->periodicTimer->Restart();
+	}
+
+	void ActiveSpeakerObserver::OnTimer(Timer* timer)
+	{
+		MS_TRACE();
+
+		Update();
+	}
+
+	void ActiveSpeakerObserver::Update()
+	{
+		MS_TRACE();
+
+		uint64_t now              = DepLibUV::GetTimeMs();
+		uint64_t levelIdleTimeout = LevelIdleTimeout - (now - this->lastLevelIdleTime);
+
+		if (levelIdleTimeout <= 0)
+		{
+			if (this->lastLevelIdleTime != 0)
+			{
+				TimeoutIdleLevels(now);
+			}
+			this->lastLevelIdleTime = now;
+		}
+
+		if (!this->mapProducerSpeaker.empty() && CalculateActiveSpeaker())
+		{
+			json data          = json::object();
+			data["producerId"] = this->dominantId;
+
+			Channel::ChannelNotifier::Emit(this->id, "dominantspeaker", data);
+		}
+	}
+
+	bool ActiveSpeakerObserver::CalculateActiveSpeaker()
+	{
+		MS_TRACE();
+
+		std::string newDominantId;
+		int32_t speakerCount = this->mapProducerSpeaker.size();
+
+		if (speakerCount == 0)
+		{
+			newDominantId = "";
+		}
+		else if (speakerCount == 1)
+		{
+			auto it = this->mapProducerSpeaker.begin();
+
+			newDominantId = it->second.producer->id;
+		}
+		else
+		{
+			Speaker* dominantSpeaker =
+			  (this->dominantId.empty()) ? nullptr : this->mapProducerSpeaker[this->dominantId].speaker;
+
+			if (dominantSpeaker == nullptr)
+			{
+				auto item       = mapProducerSpeaker.begin();
```

Comment:
```suggestion
				auto item       = this->mapProducerSpeaker.begin();
```

please, ensure that all class members (excluding class methods) use `this->`.